internal/peer/node/start.go

```diff
@@ -828,9 +834,32 @@ func serve(args []string) error {
 		}()
 	}
 
+	signalFunc := func() {
+		// notify all externally built chaincodes of signal
+		close(externalCCSignalChan)
+
+		// wait for a message from each externally built cc that
+		// has been launched that it has terminated (and cleaned
+		// up after itself) properly
+		for ccid, instance := range containerRouter.Containers() {
+			_, ok := instance.(*externalbuilder.Instance)
+			if !ok {
+				continue
+			}
+			_, started := chaincodeHandlerRegistry.Launching(ccid)
+			if started {
+				// we don't care which cc we're receiving from,
+				// as long as we receive one message per cc that
+				// has been built and launched.
+				<-externalCCDoneChan
+			}
+		}
+		serve <- nil
+	}
+
```

Comment:
This does not belong way up here. The implementation that handles the shutdown needs to live down with the handlers.